File: 9-Stack/programs/cpp/program3.cpp
// Stack implementation

#include<bits/stdc++.h>
using namespace std;

class Node{
  public:
    int data;
    Node *next;
};

class Stack{
  private:
    Node *top;
  public:
    Stack(){
      top = NULL;
    }
    void push(int x);
    int pop();
    void Display();
    int peek(int pos);
    int stackTop();
};

void Stack::push(int x){
  Node *t = new Node;
  if(t == NULL){
    cout << "Stack is full" << endl;
  }else{
    t->data = x;
    t->next = top;
    top = t;
  }
}

int Stack::pop(){
  int x = -1;
  if(top == NULL){
    cout << "Stack is Empty" << endl;
  }else{
    Node *t = top;
    top = top->next;
    x = t->data;
    delete t;
  }
  return x;
}

void Stack::Display(){
  Node *p = top;
  while(p != NULL){
    cout << p->data << " ";
    p = p->next;
  }
  cout << endl;
}

int Stack::peek(int pos){
  Node *p = top;
  for(int i = 0; p != NULL && i < pos-1; i++){
    p = p->next;
  }
  if(p != NULL){
    return p->data;
  }else{
    return -1;
  }
}

int Stack::stackTop(){
  if(top != NULL){
    return top->data;
  }else{
    return -1;
  }
}


int main(){
  Stack stk;
  stk.push(10);
  stk.push(20);
  stk.push(30);
  stk.push(40);
  stk.push(50);
  stk.Display();
  cout << stk.pop() << endl;
  stk.Display();
  cout << stk.peek(2) << endl;
  cout << stk.stackTop() << endl;
  return 0;
}

File: 9-Stack/programs/ruby/program3.rb
# Stack implementation

class Node
  attr_accessor :data, :next

  def initialize(data)
    @data = data
    @next = nil
  end
end

class Stack
  def initialize
    @top = nil
  end

  def push(x)
    t = Node.new(x)
    if t.nil?
      puts "Stack is full"
    else
      t.next = @top
      @top = t
    end
  end

  def pop
    x = -1
    if @top.nil?
      puts "Stack is Empty"
    else
      t = @top
      @top = @top.next
      x = t.data
    end
    x
  end

  def display
    p = @top
    while p
      print "#{p.data} "
      p = p.next
    end
    puts
  end

  def peek(pos)
    p = @top
    (pos - 1).times do
      p = p.next if p
    end
    p ? p.data : -1
  end

  def stack_top
    @top ? @top.data : -1
  end
end

stk = Stack.new
stk.push(10)
stk.push(20)
stk.push(30)
stk.push(40)
stk.push(50)
stk.display
puts stk.pop
stk.display
puts stk.peek(2)
puts stk.stack_top